### PR TITLE
Crm 15 endpoint customers

### PR DESCRIPTION
**Pull request by SteelWro**

### CODE REVIEW

File: backend/src/main/java/com/crm/dto/response/CustomerResponse.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+@Getter
+@Setter
+@Builder
+public class CustomerResponse {
+    private Long id;
+    private String name;
+    private String surname;
+    private String phone;
+    private String address;
+
```
Comment (daniel250592) [inline backend/src/main/java/com/crm/dto/response/CustomerResponse.java:18]: Is it generated method? Consider using annotation @EqualsAndHashCode

File: backend/src/main/java/com/crm/service/impl/CustomerServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.stream.Collectors;
+
+@Service
+@RequiredArgsConstructor
+public class CustomerServiceImpl implements CustomerService {
+
+    private final CustomerRepository repository;
+    private final CustomerMapper mapper;
+
+    @Override
```
Comment (daniel250592) [inline backend/src/main/java/com/crm/service/impl/CustomerServiceImpl.java:24]: that could be final Pageable pageable

File: .gitignore [hunk 3]
```diff
@@ -19,7 +19,6 @@
 *.zip
 *.tar.gz
 *.rar
-*.iml
```
Comment (Fellway) [inline .gitignore:22]: Why it's removed in this pr?

File: backend/src/main/java/com/crm/api/v1/CustomerController.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.web.PageableDefault;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequiredArgsConstructor
```
[the remark this answers is not in the document]
Comment (Fellway) [reply backend/src/main/java/com/crm/api/v1/CustomerController.java:16]: We don't need additional constructor for Controller class. I think currecnt implementation is correct.

File: backend/src/main/java/com/crm/api/v1/CustomerController.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+@RequestMapping("/customers")
+public class CustomerController {
+
+    private final CustomerService customerService;
+
+    @GetMapping
+    @ApiOperation(value = "Finds all customers paginated", notes = "Add param \"page\" and/or \"size\""
+            + " to specify page no. and size of each page.")
+    public ResponseEntity<Page<CustomerResponse>> getCustomersPaginated(@PageableDefault Pageable pageable) {
```
Comment (Fellway) [inline backend/src/main/java/com/crm/api/v1/CustomerController.java:25]: final

File: backend/src/main/java/com/crm/dto/mapper/CustomerMapper.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+import org.modelmapper.ModelMapper;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class CustomerMapper {
+
+    private final ModelMapper modelMapper;
+
+    public CustomerResponse convertToDto(CustomerEntity customerEntity) {
```
Comment (Fellway) [inline backend/src/main/java/com/crm/dto/mapper/CustomerMapper.java:15]: final

File: backend/src/main/java/com/crm/dto/mapper/CustomerMapper.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        return modelMapper.map(customerEntity, CustomerResponse.class);
+    }
+
+    public CustomerEntity convertToEntity(CustomerResponse customerResponse) {
```
Comment (Fellway) [inline backend/src/main/java/com/crm/dto/mapper/CustomerMapper.java:19]: final

File: backend/src/main/java/com/crm/dto/response/CustomerResponse.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+import lombok.Builder;
+import lombok.Getter;
+import lombok.Setter;
+
+import java.util.Objects;
+
+@Getter
+@Setter
+@Builder
```
Comment (Fellway) [inline backend/src/main/java/com/crm/dto/response/CustomerResponse.java:11]: Do we need builder?

File: backend/src/main/java/com/crm/model/db/CustomerEntity.java [hunk 9]
```diff
@@ -1,19 +1,20 @@
...
 import java.util.HashSet;
+import java.util.Objects;
 import java.util.Set;
 
 @Entity
 @NoArgsConstructor
 @Getter
 @Setter
 @EqualsAndHashCode
+@Builder
```
Comment (Fellway) [inline backend/src/main/java/com/crm/model/db/CustomerEntity.java:17]: Same as above

File: backend/src/main/java/com/crm/service/CustomerService.java [hunk 10]
```diff
@@ -0,0 +1,10 @@
+package com.crm.service;
+
+import com.crm.dto.response.CustomerResponse;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+
+public interface CustomerService {
+
+    Page<CustomerResponse> getCustomersPaginated(Pageable pageable);
```
Comment (Fellway) [inline backend/src/main/java/com/crm/service/CustomerService.java:9]: final

File: backend/src/main/test/java/com/crm/controller/CustomerControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,27 @@
+package java.com.crm.controller;
```
Comment (Fellway) [inline backend/src/main/test/java/com/crm/controller/CustomerControllerTest.java:1]: Remove this class

File: backend/src/main/test/java/com/crm/dto/CustomerMapperTest.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+                .id(ID)
+                .name(NAME)
+                .surname(SURNAME)
+                .phone(PHONE)
+                .address(ADDRESS)
+                .build();
+    }
+
+    @Test
+    void shouldReturnMappedCustomerToCustomerDto() {
```
Comment (Fellway) [inline backend/src/main/test/java/com/crm/dto/CustomerMapperTest.java:44]: this test, tests nothing

File: backend/src/main/test/java/com/crm/dto/CustomerMapperTest.java [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+        CustomerResponse customerResponse = mapper.convertToDto(this.customerEntity);
+
+    }
+
+    @Test
+    void shouldReturnMappedCustomerDtoToCustomer() {
```
Comment (Fellway) [inline backend/src/main/test/java/com/crm/dto/CustomerMapperTest.java:50]: this test, tests nothing

**Review by Fellway — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 3 above)
Comment (SteelWro) [reply .gitignore:22]: My fault, i do not know how it happened

File: backend/src/main/java/com/crm/api/v1/CustomerController.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (SteelWro) [reply backend/src/main/java/com/crm/api/v1/CustomerController.java:16]: Ok, so I leave it without changes

File: backend/src/main/java/com/crm/dto/mapper/CustomerMapper.java (hunk 6 above)
Comment (SteelWro) [reply backend/src/main/java/com/crm/dto/mapper/CustomerMapper.java:15]: Done

File: backend/src/main/java/com/crm/dto/response/CustomerResponse.java (hunk 8 above)
Comment (SteelWro) [reply backend/src/main/java/com/crm/dto/response/CustomerResponse.java:11]: I use @Builder in CustomerMapper Tests

File: backend/src/main/java/com/crm/service/impl/CustomerServiceImpl.java (hunk 2 above)
Comment (SteelWro) [reply backend/src/main/java/com/crm/service/impl/CustomerServiceImpl.java:24]: done

File: backend/src/main/test/java/com/crm/dto/CustomerMapperTest.java [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+                .surname(SURNAME)
+                .phone(PHONE)
+                .address(ADDRESS)
+                .build();
+    }
+
+    @Test
+    void shouldReturnMappedCustomerToCustomerDto() {
+        CustomerResponse customerResponse = mapper.convertToDto(this.customerEntity);
+
```
[the remark this answers is not in the document]
Comment (SteelWro) [reply backend/src/main/test/java/com/crm/dto/CustomerMapperTest.java:46]: Yes, my fault. I Forgot to push newest changes

File: backend/src/main/test/java/com/crm/dto/CustomerMapperTest.java [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Test
+    void shouldReturnMappedCustomerDtoToCustomer() {
+        CustomerEntity customerEntity = mapper.convertToEntity(this.customerResponse);
```
[the remark this answers is not in the document]
Comment (SteelWro) [reply backend/src/main/test/java/com/crm/dto/CustomerMapperTest.java:51]: Done

File: backend/src/main/test/java/com/crm/dto/CustomerMapperTest.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (SteelWro) [reply backend/src/main/test/java/com/crm/dto/CustomerMapperTest.java:46]: Done

File: backend/src/main/java/com/crm/api/v1/CustomerController.java (hunk 5 above)
Comment (SteelWro) [reply backend/src/main/java/com/crm/api/v1/CustomerController.java:25]: done

File: backend/src/main/java/com/crm/dto/mapper/CustomerMapper.java (hunk 7 above)
Comment (SteelWro) [reply backend/src/main/java/com/crm/dto/mapper/CustomerMapper.java:19]: done

File: backend/src/main/java/com/crm/service/CustomerService.java (hunk 10 above)
Comment (SteelWro) [reply backend/src/main/java/com/crm/service/CustomerService.java:9]: done

File: backend/src/main/test/java/com/crm/dto/CustomerMapperTest.java (hunk 12 above)
Comment (SteelWro) [reply backend/src/main/test/java/com/crm/dto/CustomerMapperTest.java:44]: fixed

File: backend/src/main/test/java/com/crm/dto/CustomerMapperTest.java (hunk 13 above)
Comment (SteelWro) [reply backend/src/main/test/java/com/crm/dto/CustomerMapperTest.java:50]: fixed

File: backend/src/main/test/java/com/crm/controller/CustomerControllerTest.java (hunk 11 above)
Comment (SteelWro) [reply backend/src/main/test/java/com/crm/controller/CustomerControllerTest.java:1]: Yes, done

File: backend/src/main/java/com/crm/model/db/CustomerEntity.java [hunk 16]
```diff
@@ -41,4 +42,16 @@
     )
     private Set<CarEntity> cars = new HashSet<>();
 
+    @Override
+    public boolean equals(Object o) {
```
Comment (Fellway) [inline backend/src/main/java/com/crm/model/db/CustomerEntity.java:46]: remove if we are using lombok annotation for equals and hashCode

File: backend/src/main/java/com/crm/model/db/CustomerEntity.java [hunk 17]
```diff
@@ -41,4 +42,16 @@
...
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        CustomerEntity that = (CustomerEntity) o;
+        return getName().equals(that.getName()) && getSurname().equals(that.getSurname()) && getPhone().equals(that.getPhone()) && getAddress().equals(that.getAddress());
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (Fellway) [inline backend/src/main/java/com/crm/model/db/CustomerEntity.java:54]: same here

**Review by Fellway — CHANGES_REQUESTED** [review]

File: backend/src/main/java/com/crm/dto/response/CustomerResponse.java (hunk 1 above)
Comment (SteelWro) [reply backend/src/main/java/com/crm/dto/response/CustomerResponse.java]: I decided to use `@EqualsAndHashCode`, thanks

File: backend/src/main/java/com/crm/model/db/CustomerEntity.java (hunk 16 above)
Comment (SteelWro) [reply backend/src/main/java/com/crm/model/db/CustomerEntity.java]: done

File: backend/src/main/java/com/crm/model/db/CustomerEntity.java (hunk 17 above)
Comment (SteelWro) [reply backend/src/main/java/com/crm/model/db/CustomerEntity.java]: done

File: backend/src/main/java/com/crm/model/db/CustomerEntity.java (hunk 9 above)
Comment (SteelWro) [reply backend/src/main/java/com/crm/model/db/CustomerEntity.java:29]: I'll leave it there

File: backend/src/main/java/com/crm/service/impl/CustomerServiceImpl.java (hunk 2 above)
Comment (SteelWro) [reply backend/src/main/java/com/crm/service/impl/CustomerServiceImpl.java:24]: done

**Review by Fellway — APPROVED** [review]

**Review by Fellway — APPROVED** [review]

**Pull request merged**